src/embed_tests/TestDomainReload.cs

```diff
@@ -50,173 +59,383 @@ class TestDomainReload
+                        using (PyObject obj = scope.Get("obj"))
+                        {
+                            Debug.Assert(obj.AsManagedObject(type).GetType() == type);
+                            // We only needs its Python handle
+                            PyRuntime.XIncref(obj.Handle);
+                            return obj.Handle;
                         }
                     }
                 }
-                static void OnDomainUnload(object sender, EventArgs e) {
-                    System.Console.WriteLine(string.Format(""[{0} in .NET] unloading"", AppDomain.CurrentDomain.FriendlyName));
+                catch (Exception e)
+                {
+                    Debug.WriteLine(e);
+                    throw;
                 }
-            }";
+            }
+        }
 
 
-        /// <summary>
-        /// Build an assembly out of the source code above.
-        ///
-        /// This creates a file <paramref name="assemblyName"/>.dll in order
-        /// to support the statement "proxy.theAssembly = assembly" below.
-        /// That statement needs a file, can't run via memory.
-        /// </summary>
-        static Assembly BuildAssembly(string assemblyName)
+        class CrossDomianObjectStep2 : CrossCaller
         {
-            var provider = CodeDomProvider.CreateProvider("CSharp");
+            public override ValueType Execte(ValueType arg)
+            {
+                // handle refering a clr object created in previous domain,
+                // it should had been deserialized and became callable agian.
+                IntPtr handle = (IntPtr)arg;
+                try
+                {
+                    using (Py.GIL())
+                    {
+                        IntPtr tp = Runtime.Runtime.PyObject_TYPE(handle);
+                        IntPtr tp_clear = Marshal.ReadIntPtr(tp, TypeOffset.tp_clear);
 
-            var compilerparams = new CompilerParameters();
-            compilerparams.ReferencedAssemblies.Add("Python.Runtime.dll");
-            compilerparams.GenerateExecutable = false;
-            compilerparams.GenerateInMemory = false;
-            compilerparams.IncludeDebugInformation = false;
-            compilerparams.OutputAssembly = assemblyName;
+                        using (PyObject obj = new PyObject(handle))
+                        {
+                            obj.InvokeMethod("Method");
+                            obj.InvokeMethod("StaticMethod");
 
-            var results = provider.CompileAssemblyFromSource(compilerparams, TestCode);
-            if (results.Errors.HasErrors)
+                            using (var scope = Py.CreateScope())
+                            {
+                                scope.Set("obj", obj);
+                                scope.Exec(@"
+obj.Method()
+obj.StaticMethod()
+obj.Property += 1
+obj.Field += 10
+");
+                            }
+                            var clrObj = obj.As<Domain.MyClass>();
```

Comment:
1. If it's not serialize, it will be skip, and you will get an exception when you use it. Even the serializable checking still need to improve: Type A contains Type B, A is serializable but B is not, only checking typeof(A).IsSerializable is not enough.
2. For user objects, I provided the RuntimeData.WrappersStorer for them. I write this to the document after this.